source/scpd/tests/VTableTest.d

```diff
@@ -0,0 +1,158 @@
+/*******************************************************************************
+
+    Contains checking if the virtual methods are in the same order
+    using vtable.
+
+    Copyright:
+        Copyright (c) 2020 BOS Platform Foundation Korea
+        All rights reserved.
+
+    License:
+        MIT License. See LICENSE for details.
+
+*******************************************************************************/
+
+module scpd.tests.VTableTest;
+
+/// In case the virtual methods are in the correct order.
+unittest
+{
+    // abstract class, this is in the correct order.
+    extern (C++) class TestA
+    {
+    public:
+        ~this()
+        {
+        }
+        abstract void vfunc1();
+        abstract void vfunc2();
+    }
+
+    extern (C++) class BonTestA : TestA
+    {
+    public:
+        override void vfunc1() {}
+        override void vfunc2() {}
+    }
+
+    uint base = 2;
+    auto n = new BonTestA();
+    // The point of the method is the same as the expected value.
+    assert(n.__vptr[base + 0] == &BonTestA.vfunc1);
+    assert(n.__vptr[base + 1] == &BonTestA.vfunc2);
+}
+
+/// In case the virtual methods are in the incorrect order.
+unittest
+{
+    // abstract class, this is in the incorrect order.
+    extern (C++) class TestB
+    {
+    public:
+        ~this()
+        {
+        }
+        abstract void vfunc2();
+        abstract void vfunc1();
+    }
+
+    extern (C++) class BonTestB : TestB
+    {
+    public:
+        override void vfunc1() {}
+        override void vfunc2() {}
+    }
+
+    uint base = 2;
+    auto n = new BonTestB();
+    // The method's point differs from the expected value.
+    assert(n.__vptr[base + 0] != &BonTestB.vfunc1);
+    assert(n.__vptr[base + 1] != &BonTestB.vfunc2);
+}
+
+/// Test the `SCPDriver`; Check that the virtual method of C++ and D are in the same order.
+unittest
+{
+    import scpd.Cpp;
+    import scpd.scp.SCPDriver;
+    import scpd.types.Stellar_types;
+    import scpd.types.Stellar_SCP;
+
+    import std.algorithm;
+    import std.range;
+    import std.traits;
+
+    import core.stdc.stdint;
+    import core.time;
+
+    string[] methods;
+
+    foreach (member; __traits(allMembers, SCPDriver))
+    {
+        if ((member == "__dtor") || (member == "__xdtor")) continue;
```

Comment:
Why ? dtors are also put in the vtable

Reply:
I've learned some new facts.

1. Taking the address of a destructor is forbidden
    However, the address of a destructor is the second value of the `vtable`.
2.
In C++, the virtual methods were ordered in the following ways:
However, if the destructor is located in the middle, not the first, then the number is confused.
&SCPDriver::signEnvelope
I need to find a new way to get the address of the virtual method vtable.
I need more correction to solve this problem.
I am going to modify getMethodSeqSCPDriver.

Reply:
https://github.com/bpfkorea/agora/pull/689

Reply:
> In C++, the virtual methods were ordered in the following ways:


Looks like something is missing after this sentence ?

> However, the address of a destructor is the second value of the vtable.

C++ generates two destructors on most compilers. See [this SO question](https://stackoverflow.com/questions/44558119/why-do-i-have-two-destructor-implementations-in-my-assembly-output).

Reply:
I saw two of them with the same address. The two addresses were `__dtor`.

Reply:
And I find that the addresses of `__dtor` are always `__vptr[0]` and `__vptr[1]`.